rfcs/012-add-session-ssh-support/012-add-session-ssh-support.md

```diff
@@ -0,0 +1,438 @@
+- Start Date: 2023-01-20
+- Status: Proposed
+
+# Add ssh support to sessions
+
+## Summary
+
+> One paragraph explanation of the change.
+
+An important requirement which must be met by Renku is to enable users to
+log in to their Renku session via ssh; a key driver for this is to support
+working with VS Code within Renku sessions.
+
+This has been discussed within the team and an interim solution has been
+developed which is currently in operation. That solution, however, has some
+limitations; this RFC provides context, documents possible solutions and will
+be used to justify the final agreed solution.
+
+## Motivation
+
+> Why are we doing this? What use cases does it support? What is the expected
+outcome?
+
+There is a [Shape-up
+document](https://www.notion.so/Support-RenkuLab-compute-access-from-local-terminal-SSH-VSCode-f896d3b391c94bcc87c56e375eb531d6)
+which provides the motivation for introducing this functionality.
+
+## Problem Definition
+
+> This section should include a detailed description of the problem and, importantly,
+include any specific constraints that arise. It can be as technical as required.
+
+The basic problem is to provide low friction, secure access to user sessions
+via ssh. 
+
+Although non key based mechansisms are possible and are discussed to some extent
+below, VS Code primarily uses key based mechansims; they are generally the most
+widely used with ssh and are known to be trusted and secure - for this reason,
+there is a strong focus on key-based mechanisms here.
+
+From a user perspective, the following needs to be considered:
+- key setup
```

Comment:
can we simply insist on being in control of the keys? If the setup is automatic I don't see much of a need for supporting all the possible options. afaik other solutions of this type (e.g. gitpod, codespaces) don't give you any options regarding this and I never felt like this was limiting somehow. I think we _have_ to do the automatic setup anyway and then it might just save us a lot of time (at least in the first stab at this) to call it good at that point.

Reply:
I was just trying to itemize the different possibilities I see; definitely agree that it makes sense not to bite off all at once.

Regarding us being in control of the keys - I guess you simply mean that renku generates a keypair and the private key stays with the user.

My pref would be for a default behaviour which simply checks if a renku key exists and generates one if not, but if a flag/switch is provided, an a priori generated key can be used.

Reply:
sure, that makes sense. On the spectrum from "required" to "not at all needed" I would then make the automatic key generation "required" and the others "nice-to-have" (i.e. if easy or with enough time left, do it on the first implementation, otherwise defer to subsequent improvements.